ec2/instance.py

```diff
@@ -14,165 +14,208 @@
+        a runtime property.
+        If use_external_resource is false or not given, the image_id
+        and instance_type properties are taken from the node properties
+        and inserted into the run_instances boto function arguments.
+        Then the rest of the parameters are added to the arguments,
+        overriding image_id and instance_type, if given.
+        Except min_count and max_count, which are always set to 1.
+        The run_instance function is sent to EC2. If no error is raised,
+        the output of that function is returned and the instance id is
+        assigned to the runtime properties.
     """
+
     ec2_client = connection.EC2ConnectionClient().client()
 
-    arguments = dict()
-    arguments['image_id'] = ctx.node.properties['image_id']
-    arguments['instance_type'] = ctx.node.properties['instance_type']
-    args_to_merge = build_arg_dict(ctx.node.properties['attributes'].copy(),
-                                   RUN_INSTANCES_UNSUPPORTED)
-    arguments.update(args_to_merge)
-
-    ctx.logger.info('Creating EC2 Instance.')
-    ctx.logger.debug('Attempting to create EC2 Instance.'
-                     'Image id: {0}. Instance type: {1}.'
-                     .format(arguments['image_id'],
-                             arguments['instance_type']))
-    ctx.logger.debug('Sending these API parameters: {0}.'
-                     .format(arguments))
+    if ctx.node.properties['use_external_resource']:
+        instance_id = ctx.node.properties['resource_id']
+        instance = utils.get_instance_from_id(instance_id, ctx=ctx)
+        ctx.instance.runtime_properties['aws_resource_id'] = instance.id
+        ctx.logger.info('Using existing instance: {0}.'.format(instance.id))
+        return
+
+    instance_parameters = utils.get_instance_parameters(ctx=ctx)
+
+    ctx.logger.debug(
+        'Attempting to create EC2 Instance. Sending these API parameters: {0}.'
+        .format(instance_parameters))
 
     try:
-        reservation = ec2_client.run_instances(**arguments)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('Error. Failed to run EC2 Instance: '
-                                  'API returned: {0}.'.format(e))
+        reservation = ec2_client.run_instances(**instance_parameters)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('{0}'.format(str(e)))
 
     instance_id = reservation.instances[0].id
-    ctx.instance.runtime_properties['instance_id'] = instance_id
-
-    if utils.validate_instance_id(reservation.instances[0].id, ctx=ctx):
-        utils.validate_state(reservation.instances[0], INSTANCE_RUNNING,
-                             CREATION_TIMEOUT, CHECK_INTERVAL, ctx=ctx)
+    ctx.instance.runtime_properties['aws_resource_id'] = instance_id
+    ctx.logger.info('Created instance: {0}.'.format(instance_id))
 
```

Comment:
check that id exists rather than doing the 'does not exist' thing in the start call down below